packages/eslint-plugin/typings/ts-eslint.d.ts

```diff
@@ -0,0 +1,416 @@
+  interface ReportDescriptor<TMessageIds extends string> {
+    /**
+     * The parameters for the message string associated with `messageId`.
+     */
+    data?: Record<string, any>;
+    /**
+     * The fixer function.
+     */
+    fix?: ReportFixFunction | null;
+    /**
+     * The messageId which is being reported.
+     */
+    messageId: TMessageIds;
+    /**
+     * The Node or AST Token which the report is being attached to
+     */
+    node: TSESTree.Node | AST.Token;
+    /**
+     * An override of the location of the report
+     */
+    loc?: TSESTree.SourceLocation;
+  }
+
+  interface RuleContext<TMessageIds extends string, TOptions extends any[]> {
+    /**
+     * The rule ID.
+     */
+    id: string;
+    /**
+     * An array of the configured options for this rule.
+     * This array does not include the rule severity.
+     */
+    options: TOptions;
+    /**
+     * The shared settings from configuration.
+     * We do not have any shared settings in this plugin.
+     */
+    settings: {};
+    /**
+     * The name of the parser from configuration.
+     */
+    parserPath: string;
+    /**
+     * The parser options configured for this run
+     */
+    parserOptions: Linter.ParserOptions;
+    /**
+     * An object containing parser-provided services for rules
+     */
+    parserServices?: ParserServices;
+
+    /**
+     * Returns an array of the ancestors of the currently-traversed node, starting at
+     * the root of the AST and continuing through the direct parent of the current node.
+     * This array does not include the currently-traversed node itself.
+     */
+    getAncestors(): TSESTree.Node[];
+
+    /**
+     * Returns a list of variables declared by the given node.
+     * This information can be used to track references to variables.
+     */
+    getDeclaredVariables(node: TSESTree.Node): Scope.Variable[];
+
+    /**
+     * Returns the filename associated with the source.
+     */
+    getFilename(): string;
+
+    /**
+     * Returns the scope of the currently-traversed node.
+     * This information can be used track references to variables.
+     */
+    getScope(): Scope.Scope;
+
+    /**
+     * Returns a SourceCode object that you can use to work with the source that
+     * was passed to ESLint.
+     */
+    getSourceCode(): SourceCode;
+
+    /**
+     * Marks a variable with the given name in the current scope as used.
+     * This affects the no-unused-vars rule.
+     */
+    markVariableAsUsed(name: string): boolean;
+
+    /**
+     * Reports a problem in the code.
+     */
+    report(descriptor: ReportDescriptor<TMessageIds>): void;
+  }
+
+  // This isn't the correct signature, but it makes it easier to do custom unions within reusable listneers
+  // never will break someone's code unless they specifically type the function argument
+  type RuleListener = Record<string, (node: never) => void>;
+
+  interface RuleModule<
+    TMessageIds extends string,
+    TOptions extends any[],
```

Comment:
you may consider making those generics optional
```
    TMessageIds extends string = never,
    TOptions extends any[] = never,
```
in case if there is no options provided

Reply:
I thought about that, but I decided to leave it.

I thought that explicitly setting the generic to `[]` at call time is probably clearer to readers of the rule:

```TS
const rule: RuleModule<MessageIds, []> = {
// versus
const rule: RuleModule<MessageIds> = {
```